pulsar/producer_partition.go

```diff
@@ -236,8 +236,18 @@ func (p *partitionProducer) ConnectionClosed() {
 }
 
 func (p *partitionProducer) reconnectToBroker() {
-	backoff := internal.Backoff{}
-	for {
+	var (
+		maxRetry int
+		backoff  = internal.Backoff{}
+	)
+
+	if p.options.MaxReconnectToBroker == nil {
+		maxRetry = -1
+	} else {
+		maxRetry = int(*p.options.MaxReconnectToBroker)
+	}
```

Comment:
The handling for uninitialized maxRetry seems not correct, it may be better to extract the retry test into a separate function, e.g `shouldRetry`.

Reply:
> The handling for uninitialized maxRetry seems not correct, it may be better to extract the retry test into a separate function, e.g `shouldRetry`.

```go
func (p *partitionProducer) limitRetry() bool {
 return p.options.MaxReconnectToBroker != nil
}

func (p *partitionProducer) reconnectToBroker() {
...
 var maxRetry = -1
 if p.limitRetry() {
   maxRetry = int(*p.options.MaxReconnectToBroker)  
 }
...
}
```

Do you mean something like this?

Reply:
my bad, the handling logic is correct :-) 

What i mean is to wrap the test logic, like following

```go
func (p *partitionProducer) shouldRetry() bool {
	maxReconnects := p.opts.MaxReconnectToBroker

	switch {
	case maxReconnects < 0:
		return true
	case maxReconnects == 0:
		return false
	default:
		p.reconnectAttempts++
		if p.reconnectAttempts > maxReconnects {
			return false
		}

		return true
	}
}

func (p *partitionProducer) reconnectToBroker() {
	for p.shouldRetry() {
		// retry logic
	}
}
```

Reply:
Great, This looks more clear. 
I think it would be better if we decide the type of p.opts.MaxReconnectToBroker first.
Besides, i think `shouldRetry()` logic maybe a little bit slower than the origin one, we can reduce the `shouldRetry()` call times by move some common logic outside of the for loop.